lib/smack/NormalizeLoops.cpp

```diff
@@ -0,0 +1,134 @@
+//
+// This file is distributed under the MIT License. See LICENSE for details.
+//
+
+//
+// This pass normalizes loop structures to allow easier generation of Boogie
+// code when a loop edge comes from a branch. Specifically, a forwarding block
+// is added between a branching block and the loop header.
+//
+
+#include "smack/NormalizeLoops.h"
+#include "smack/Naming.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/ValueSymbolTable.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/Analysis/LoopInfo.h"
+#include "llvm/IR/Dominators.h"
+#include <map>
+#include <set>
+
+namespace smack {
+
+using namespace llvm;
+
+// Register LoopInfo
+void NormalizeLoops::getAnalysisUsage(AnalysisUsage& AU) const {
+  AU.addRequired<LoopInfoWrapperPass>();
+}
+
+/**
+ * Creates a basic block in \param func which unconditionally branches to
+ * \param target.
+ */
+BasicBlock* makeForwardingBlock(BasicBlock* target, Function* func) {
+  auto& context = target->getContext();
+  BasicBlock* result = BasicBlock::Create(context, "forwarder", func);
+  BranchInst* branch = BranchInst::Create(target);
+  result->getInstList().push_back(branch);
+  return result;
+}
+
+/**
+ * Splits any edge between \param BB and \param headerBlock in a loop body,
+ * by adding a forwarding block between \param BB and \param headerBlock. The
+ * \param blockMap is updated to map between \param BB and the newly created
+ * forwarding block.
+ */
+void splitBlock(BasicBlock* BB, BasicBlock* headerBlock, Function* func,
+                std::map<BasicBlock*, BasicBlock*>& blockMap) {
+  TerminatorInst* ti = BB->getTerminator();
+  unsigned numSuccs = ti->getNumSuccessors();
+  if (numSuccs <= 1) { return; }
+  BasicBlock* forwarder = nullptr;
+  for (unsigned i = 0; i < numSuccs; ++i) {
+    BasicBlock* succ = ti->getSuccessor(i);
+    if (succ != headerBlock) { continue; }
+
+    // Successor is the header block and this block has multiple successors,
+    // so it needs a forwarding block.
+    if (forwarder == nullptr) {
+      forwarder = makeForwardingBlock(headerBlock, func);
+      blockMap[BB] = forwarder;
+    }
+    ti->setSuccessor(i, forwarder);
+  }
+}
+
+/**
+ * \param blockMap a map from old BasicBlocks and new BasicBlocks.
+ * Fixes any PHI nodes in \param phis to get their incoming block from the new
+ * blocks in \param blockMap.
+ */
+void processPhis(std::vector<PHINode*>& phis, std::map<BasicBlock*, BasicBlock*>& blockMap) {
+  for (PHINode* phi : phis) {
+    unsigned numIncoming = phi->getNumIncomingValues();
+    for (unsigned i = 0; i < numIncoming; ++i) {
+      BasicBlock* incomingBlock = phi->getIncomingBlock(i);
+      if (blockMap.count(incomingBlock)) {
+        phi->setIncomingBlock(i, blockMap[incomingBlock]);
+      }
+    }
+  }
+}
+
+void processLoop(Loop *loop, Function* func) {
```

Comment:
There’s no need for `func` to be a parameter here or in `splitBlock` or `makeForwardingBlock`, since it can be determined when it is needed in `makeForwardingBlock` as `target->getParent()`.

Reply:
Done.